### ISSUE TITLE
DB refactor positions/attributes

### ISSUE THREAD

**Issue by tsmgeek** [tananaev/traccar#3442]
Im noticing that the positions table can grow at an alarming rate if you have a large amount of trackers logging quickly also if they return lots of network data it is made worse.
Would it not be an idea to try and migrate the network data into its own table and reference it from the positions table via a map table?  (positions > map > networks).

This will have advantages, where needing to return lots of tracks for the most part the network data is not really required, this could be left out and as it is in a separate table this will be also beneficial for the DB.

The same could be said of the attributes table, possibly make this into a KV.

**Comment by tananaev** [comment]
The easiest solution is to have a configuration flag to not save network data, if user is not interested in it.

**Comment by tsmgeek** [comment]
One way yes, but really as it does not sort the root problem out.
Data normalisation, there is lots of data that is repeated and just fills the logs up for no reason. Attributes is worst, for the most part things dont change, would be better if they were not logged or logged only when they do change.

**Comment by tananaev** [comment]
You just described two completely different problems.

**Comment by tsmgeek** [comment]
Not really, its still all about reducing how much duplicate data is being stored.

**Comment by tananaev** [comment]
Let's start by analysing if there is a problem. Can you please provide some numbers.

**Comment by tsmgeek** [comment]
Lets assume the following which are based on typical rows from my two trackers.

Network: ~145 bytes (more if device is logging multiple cell towers or wifi aps)
Attributes: ~369 bytes (in my case, 2xADC, GPIOx4, IP etc)
Address: ~35 bytes (this can vary quite a bit)
Proto: 6 bytes (in my case)

Total size of variable length fields is 555bytes (10s sample intervals while moving, 4795200b per day), this is assuming the device is moving all day long, even if we say 3h per day that is still 599400b, it can build up quickly.

Assume someone had lots of trackers working all day long, depending on the update frequency it would be quite high.

- All protos could be stored in a separate table and use a byte/short to reference it, load a protomap on init and reference it like that.
- Addresses table mapped, its possable that many trackers will sit in the same location when parked up etc.

The ultimate would be to use systems like Mongo to store attribute/network data as that has native data types that can better store that data.

**Comment by tananaev** [comment]
I think you are trying to solve wrong problems here. This is the first time in 10 years I hear concern about database size. Storage is extremely cheap these days, so I don't think it's worth it introducing more complications into the code just to save a few bytes.

**Comment by tsmgeek** [comment]
This solves several problems. Just because storage is cheap does not mean database design should be taken for granted, DRY is a valid point here.
Try and do an alter on a table which has all of this additional text data and it will take ages, but on a table which is free of this it is far quicker and easier.
Sorts etc where that data is not really required may end up involving temp tables which in turn puts pressure on disk.
Easier to return data from the API without the network data or attributes, would speed up loading large routes.

**Comment by tananaev** [comment]
I agree that it can be improve, but there are much bigger architectural problems in Traccar that need solving.

**Comment by tananaev** [comment]
Closing as no interest.
